### DISCUSSION TITLE
Question: Compilation time

### DISCUSSION THREAD

**Discussion by renlite** [c3lang/c3c#1180] [General]
Is there an experience about compilation time?
Does C3 use IC?
Is C3 compareable with Go concerning compilation speed?

**Comment by lerno** [comment]
C3's compilation time depends very much on LLVM. The LLVM part is over 95% of the compilation time. So if you have something which has an error, you'll get the error quickly. The LLVM compilation is multithreaded, so you'll get an advantage on multicore systems. As a rule of thumb, it should compile faster than Clang compiles pure C code.

**Comment by renlite** [comment]
Is IC implemented? Code changes in a single file/module should not compile the whole code base. Is this supported in C3 + LLVM process?

**Reply by lerno** [reply]
If you by "IC" mean incremental compilation, then the answer is "currently no but can be supported".

To explain in some detail:
C3 supports both separate and single module compilation. Without optimization turned on, the default is *separate compilation*. Separate compilation of course being a prerequisite for incremental compilation.

However, the C3 compiler *does not* track compilation state between compilations when using the built-in project compiler and will rebuild all files from scratch every time. In addition it does *prune unused code* before emitting a module.

It is theoretically possible to use something like Make to handle dependency management. In this case, unused code pruning should **not** be enabled. In addition the compiler probably needs to support suppressing emits for particular modules, which is not available yet. If you wish to set up such an incremental workflow with Make, please let me know.

However, at this point in time there are few projects that would cause compilation to exceed 1-2 seconds of compilation time, so it's not a problem which is highly prioritized at this point in time. Once there are projects that have more than 10s of compilation time from scratch then it's possible to actually look at optimizing it. Right now with sub-second compilation times it's really hard to optimize – the workloads are simply too small.

**Reply by renlite** [reply]
Yes, IC for incremental compilation. Thanks, for the details.
